rotkehlchen/exchanges/bitstamp.py

```diff
@@ -0,0 +1,711 @@
+            fee=deserialize_fee(raw_trade['fee']),
+            fee_currency=fee_currency,
+            link=str(raw_trade['id']),
+            notes='',
+        )
+        return trade
+
+    @staticmethod
+    def _get_trade_pair_data_from_transaction(raw_result: Dict[str, Any]) -> TradePairData:
+        """Given a user transaction that contains the base and quote assets'
+        symbol as keys, return the Bitstamp trade pair data (raw pair str,
+        base/quote assets raw symbols, and TradePair).
+
+        NB: any custom pair conversion (e.g. from Bitstamp asset symbol to world)
+        should happen here.
+
+        Can raise DeserializationError.
+        """
+        try:
+            pair = [key for key in raw_result.keys() if '_' in key and key != 'order_id'][0]
+        except IndexError:
+            raise DeserializationError(
+                'Could not deserialize Bitstamp trade pair from user transaction. '
+                f'Trade pair not found in: {raw_result}.',
+            )
+
+        # NB: `pair_get_assets()` is not used for simplifying the calls and
+        # storing the raw pair strings.
+        base_asset_symbol, quote_asset_symbol = pair.split('_')
+        try:
+            base_asset = Asset(base_asset_symbol)
+            quote_asset = Asset(quote_asset_symbol)
+        except UnknownAsset as e:
+            log.error(str(e))
+            raise DeserializationError(
+                f'Unknown asset {e.asset_name} found while processing trade pair.',
+            )
+
+        return TradePairData(
+            pair=pair,
+            base_asset_symbol=base_asset_symbol,
+            quote_asset_symbol=quote_asset_symbol,
+            base_asset=base_asset,
+            quote_asset=quote_asset,
+            trade_pair=TradePair(f'{base_asset.identifier}_{quote_asset.identifier}'),
+        )
+
+    @overload  # noqa: F811
+    def _process_unsuccessful_response(  # pylint: disable=no-self-use
+            self,
+            response: Response,
+            case: Literal['validate_api_key'],
+    ) -> Tuple[bool, str]:
+        ...
+
+    @overload  # noqa: F811
+    def _process_unsuccessful_response(  # pylint: disable=no-self-use
+            self,
+            response: Response,
+            case: Literal['balances'],
+    ) -> Tuple[Optional[Dict[Asset, Balance]], str]:
+        ...
+
+    @overload  # noqa: F811
+    def _process_unsuccessful_response(  # pylint: disable=no-self-use
+            self,
+            response: Response,
+            case: Literal['trades'],
+    ) -> List[Trade]:
+        ...
+
+    @overload  # noqa: F811
+    def _process_unsuccessful_response(  # pylint: disable=no-self-use
+            self,
+            response: Response,
+            case: Literal['asset_movements'],
+    ) -> List[AssetMovement]:
+        ...
+
+    def _process_unsuccessful_response(
+            self,
+            response: Response,
+            case: str,
+    ) -> Union[
```

Comment:
Just a note (don't change anything): Not a big fan of the amount of different possible returns of this function.